### PR TITLE
adding vpc filter for dhcp-options

### PR DESCRIPTION
**Pull request by joshuaroot**
- per #1616 - adding ability to filter vpc based on dhcp-options

### CODE REVIEW

**Comment by joshuaroot** [comment]
i'm not a huge fan of that long `if` tree... but i couldn't think of a better way to handle it...

File: c7n/resources/vpc.py [hunk 1]
```diff
@@ -241,6 +241,80 @@ def process(self, resources, event=None):
...
+
+    """
+    schema = type_schema(
+        'dhcp-options',
+        domainname={'type': 'array', 'items': {'type': 'string'}},
+        nameserver={'type': 'array', 'items': {'type': 'string'}},
+        ntpserver={'type': 'array', 'items': {'type': 'string'}})
+    permissions = ('ec2:DescribeDhcpOptions',)
+
+    def validate(self):
```
Comment (kapilt) [inline c7n/resources/vpc.py]: this entire validate method could be replaced by a better jsonschema definition.

File: c7n/resources/vpc.py [hunk 2]
```diff
@@ -241,6 +241,80 @@ def process(self, resources, event=None):
...
+
+        if ('domainname' in self.data and
+                not isinstance(self.data.get('domainname'), list)):
+            raise FilterValidationError('domainname must be a list')
+        if ('nameserver' in self.data and
+                not isinstance(self.data.get('nameserver'), list)):
+            raise FilterValidationError('nameserver must be a list')
+        if ('ntpserver' in self.data and
+                not isinstance(self.data.get('ntpserver'), list)):
+            raise FilterValidationError('ntpserver must be a list')
```
Comment (kapilt) [inline c7n/resources/vpc.py]: also fwiw validate methods return self if successful.

File: c7n/resources/vpc.py [hunk 3]
```diff
@@ -241,6 +241,80 @@ def process(self, resources, event=None):
...
+
+    def process(self, resources, event=None):
+        dmn = self.data.get('domainname', [])
+        dns = self.data.get('nameserver', [])
+        ntp = self.data.get('ntpserver', [])
+        client = local_session(self.manager.session_factory).client('ec2')
+        results = []
+        for r in resources:
+            options = client.describe_dhcp_options(
```
Comment (kapilt) [inline c7n/resources/vpc.py]: i don't think its worthwhile given the low cardinality of vpcs, but dhcps options maybe common across vpcs, also more efficient to bulk query for them, ie. make mapping of dhcp-id -> vpc_ids and dhcp_id -> options.

File: c7n/resources/vpc.py [hunk 4]
```diff
@@ -241,6 +241,80 @@ def process(self, resources, event=None):
...
+            d_names, n_svr, t_svr = ([] for i in range(3))
+            for o in options:
+                if o['Key'] == 'domain-name':
+                    d_names = [v['Value'] for v in o['Values']]
+                if o['Key'] == 'domain-name-servers':
+                    n_svr = [v['Value'] for v in o['Values']]
+                if o['Key'] == 'ntp-servers':
+                    t_svr = [v['Value'] for v in o['Values']]
+
+            if dmn and dns and ntp:
```
Comment (kapilt) [inline c7n/resources/vpc.py]: this feels like it could be rewritten to be a bit simpler.

File: c7n/resources/vpc.py [hunk 5]
```diff
@@ -241,6 +241,99 @@ def process(self, resources, event=None):
...
+    def process_vpc(self, r):
+        def _breakout(opt):
+            if opt:
+                return opt[0]
+            return []
+
+        client = local_session(self.manager.session_factory).client('ec2')
+        opts = client.describe_dhcp_options(
+            DhcpOptionsIds=[r['DhcpOptionsId']])['DhcpOptions'][0]
+        r['DhcpConfigurations'] = {
```
Comment (kapilt) [inline c7n/resources/vpc.py]: annotations need a c7n prefix

File: c7n/resources/vpc.py [hunk 6]
```diff
@@ -241,6 +241,92 @@ def process(self, resources, event=None):
...
+
+    def process(self, resources, event=None):
+        self.dmn = self.data.get('domainname', [])
+        self.dns = self.data.get('nameserver', [])
+        self.ntp = self.data.get('ntpserver', [])
+        results = []
+        for r in resources:
+            r = self.process_vpc(r)
+            if (self.dmn == r['c7n:DhcpConfigurations']['domainname'] and
+                    self.dns == r['c7n:DhcpConfigurations']['nameserver'] and
```
Comment (kapilt) [inline c7n/resources/vpc.py]: this matching leaves a lot to be desired, the same two sets of nameservers but in different order wouldn't match.

**Pull request closed without merging**